File: src/main/services/page-count.ts
import { existsSync } from 'fs'
import { countCbzPages } from './apply-metadata'
import { getPdfPageCount } from './pdf-extract'

/**
 * How many pages a library file holds, whatever its format.
 *
 * One place, because the answer is needed at every point a file is written —
 * download, conversion, sync — and by the detail and series panels. Two
 * formats, two mechanisms:
 *
 *   cbz  image entries in the archive's central directory, nothing inflated
 *   pdf  `pdfinfo`, which reads the trailer rather than the document
 *
 * Both are metadata reads and cost around 25ms on a network mount. That is
 * cheap once and far too expensive on every render, which is why the result is
 * stored on the row rather than recomputed.
 *
 * Returns null rather than throwing. A page count is an enrichment: a missing
 * or unreadable file, or a PDF on a machine without poppler, should leave the
 * field blank, never fail the download or conversion that asked for it.
 */
export async function countPages(
  filePath: string,
  format: string | null | undefined
): Promise<number | null> {
  if (!filePath || !existsSync(filePath)) return null

  const kind =
    (format || '').toLowerCase() || (filePath.toLowerCase().endsWith('.pdf') ? 'pdf' : 'cbz')

  try {
    const pages = kind === 'pdf' ? await getPdfPageCount(filePath) : await countCbzPages(filePath)
    return Number.isFinite(pages) && pages > 0 ? pages : null
  } catch {
    return null
  }
}

File: kopibon-core/src/download/page_count.rs
//! page-count.ts port: how many pages a file holds, whatever its format.
//! cbz: image entries in the central directory, nothing inflated; pdf: 1.x
//! shells to `pdfinfo` — the port reads the trailer via lopdf (same answer,
//! no poppler dependency; 03 §6 note). Returns None rather than throwing —
//! a page count is an enrichment, never a failed download.

use crate::metadata::writers::comicinfo::count_cbz_pages;

pub fn count_pages(file_path: &std::path::Path, format: Option<&str>) -> Option<i64> {
    if !file_path.exists() {
        return None;
    }

    let kind = match format.map(|f| f.to_lowercase()) {
        Some(f) if !f.is_empty() => f,
        _ => {
            if file_path.to_string_lossy().to_lowercase().ends_with(".pdf") {
                "pdf".to_string()
            } else {
                "cbz".to_string()
            }
        }
    };

    let pages = match kind.as_str() {
        "pdf" => pdf_page_count(file_path),
        _ => count_cbz_pages(file_path).ok(),
    };
    match pages {
        Some(n) if n > 0 => Some(n as i64),
        _ => None,
    }
}

/// getPdfPageCount equivalent: /Count from the page tree (pdfinfo parses the
/// same trailer): Root (catalog) → Pages → Count.
fn pdf_page_count(path: &std::path::Path) -> Option<usize> {
    let doc = lopdf::Document::load(path).ok()?;
    let catalog_id = doc.trailer.get(b"Root").ok()?.as_reference().ok()?;
    let catalog = doc.get_object(catalog_id).ok()?.as_dict().ok()?;
    let pages_id = catalog.get(b"Pages").ok()?.as_reference().ok()?;
    let pages = doc.get_object(pages_id).ok()?.as_dict().ok()?;
    pages.get(b"Count").ok()?.as_i64().ok().map(|n| n as usize)
}
